### PR TITLE
allow running as arbitrary uid

### LINKED ISSUE #147: Make UID & GID configurable
## Expected Behavior
It should be possible to set the uid or gid for the `couchdb` (linux) user using environment variables.

## Current Behavior
The uid and gid are hardcoded to the default port `5984` as in #53 decided.

## Possible Solution
Use the environment variables `PUID` and `PGID` to set them.

## Context
The downside of the decision from #53, implemented in #64, is, that mounted volumes including config files are harder to edit when logged in on the host system with another uid like `1000`. Prefixing them with a `P` avoids confusion with ids of the current user.
[LinuxServer.io wrote about this](https://www.linuxserver.io/docs/puid-pgid). They create a user and group `abc` with the id `911` and change it with [this script](https://github.com/linuxserver/docker-baseimage-ubuntu/blob/master/root/etc/cont-init.d/10-adduser) in their base images.

### PR DESCRIPTION
**Pull request by willholley**
## Overview

Enables the CouchDB container to run as a non-root / user-specified uid. For now, I've only modified the 2.3.1 container but if the approach is accepted we can easily backport it.

Specifically, this:
 * Adds guards around entrypoints commands that require root (as suggested in #147)
 * Broadens permissions within the container filesystem to allow access by non-couchdb users.
 * Adds an example to the documentation which specifies `--user`.

## Testing recommendations

Build the container locally:
```
docker build -t couchdb:2.3.1 2.3.1
```

Run the container as a custom user
```
docker run --user <user> --name mycouch -p 5984:5984 -d couchdb:2.3.1
```

or, run the container as a custom user with a mounted data volume:
```
mkdir /Users/<user>/temp/couchdb-data
docker run --user <user> --name mycouch -v /Users/<user>/temp/couchdb-data:/opt/couchdb/data -p 5984:5984 -d couchdb:2.3.1
```

Verify Couch is working:
```
curl localhost:5984
```

## GitHub issue number

Fixes #147 

## Checklist

- [ ] Code is written and works correctly;
- [ ] Changes are covered by tests;
- [ ] Documentation reflects the changes;

### CODE REVIEW

**Comment by wohali** [comment]
Hey @willholley , will review this later today. Looks like `dev` is failing for other reasons, making Travis fail. Do you have time to lend a hand?

File: 2.3.1/docker-entrypoint.sh [hunk 1]
```diff
@@ -25,36 +25,44 @@ if [ "$1" = 'couchdb' ]; then
...
 
-	# Ensure that data files have the correct permissions. We were previously
-	# preventing any access to these files outside of couchdb:couchdb, but it
-	# turns out that CouchDB itself does not set such restrictive permissions
-	# when it creates the files. The approach taken here ensures that the
-	# contents of the datadir have the same permissions as they had when they
-	# were initially created. This should minimize any startup delay.
-	find /opt/couchdb/data -type d ! -perm 0755 -exec chmod -f 0755 '{}' +
-	find /opt/couchdb/data -type f ! -perm 0644 -exec chmod -f 0644 '{}' +
+		# Ensure that data files have the correct permissions. We were previously
```
Comment (wohali) [inline 2.3.1/docker-entrypoint.sh]: Fix tabs please

**Comment by wohali** [comment]
@kocolosk you've been doing the reviews in this space, care to have a look?

**Comment by kocolosk** [comment]
Will do

File: 2.3.1/Dockerfile [hunk 2]
```diff
@@ -121,8 +121,13 @@ COPY docker-entrypoint.sh /usr/local/bin
 RUN ln -s usr/local/bin/docker-entrypoint.sh /docker-entrypoint.sh # backwards compat
 ENTRYPOINT ["tini", "--", "/docker-entrypoint.sh"]
 
-# Setup directories and permissions
-RUN find /opt/couchdb \! \( -user couchdb -group couchdb \) -exec chown -f couchdb:couchdb '{}' +
```
Comment (kocolosk) [inline 2.3.1/Dockerfile]: I'm worried that dropping this line this will cause a regression of #131 when starting as root and running as the CouchDB user.

Comment (willholley) [reply 2.3.1/Dockerfile]: I didn't notice any issue with startup times when testing but have added this line back for consistency.

File: 2.3.1/docker-entrypoint.sh (hunk 1 above)
Comment (willholley) [reply 2.3.1/docker-entrypoint.sh]: the existing Dockerfile uses a mix of tabs and spaces - what's the preferred option? :)

File: 2.3.1/Dockerfile (hunk 2 above)
Comment (kocolosk) [reply 2.3.1/Dockerfile]: I couldn't repro the original issue myself, either. I think it mostly showed up when running with relatively slow PVs underneath.

**Comment by willholley** [comment]
> Hey @willholley , will review this later today. Looks like `dev` is failing for other reasons, making Travis fail. Do you have time to lend a hand?

hopefully this is fixed in #152

File: 2.3.1/docker-entrypoint.sh (hunk 1 above)
Comment (wohali) [reply 2.3.1/docker-entrypoint.sh]: Oops ;) Any tabs are mistakes.

The standard I've tried to keep everywhere is: 2 or 4 space indent, spaces only.

Unless it's a Makefile, of course.

File: 2.3.1/Dockerfile (hunk 2 above)
Comment (wohali) [reply 2.3.1/Dockerfile]: Yeah, specifically those mounted in network file systems (NFS or slow SAN).

**Comment by wohali** [comment]
@willholley I'd like to merge this but @kocolosk 's comment is still unaddressed. We need to not reintroduce a regression here. Can you help? Thanks.

File: 2.3.1/Dockerfile (hunk 2 above)
Comment (willholley) [reply 2.3.1/Dockerfile]: Ok - I've added this line back at https://github.com/apache/couchdb-docker/pull/151/commits/f9a639120bda80e3b252b089cd50aa33483ed148#diff-ff57d65ca713e117d90559824747793aR127

**Comment by willholley** [comment]
@wohali I added back the chmod line in the Dockerfile which I think addresses @kocolosk's comments. I've also normalised the whitespace to use 4 spaces everywhere.

**Comment by wohali** [comment]
@willholley merge at your convenience!

I'll need to update the official docs upstream with your change, too, once this lands.

File: 2.3.1/Dockerfile [hunk 3]
```diff
@@ -121,8 +121,17 @@ COPY docker-entrypoint.sh /usr/local/bin
...
+
+RUN set -xe; \
+# Check we own everything in /opt/couchdb. Matches the command in dockerfile_entrypoint.sh
+    find /opt/couchdb \! \( -user couchdb -group couchdb \) -exec chown -f couchdb:couchdb '{}' +; \
+# Setup directories and permissions for config. Technically these could be 555 and 444 respectively
+# but we keep them as 755 and 644 for consistency with CouchDB defaults and the dockerfile_entrypoint.sh.
+    find /opt/couchdb/etc -type d ! -perm 0755 -exec chmod -f 0755 '{}' +; \
+    find /opt/couchdb/etc -type f ! -perm 0644 -exec chmod -f 0644 '{}' +; \
+# only local.d needs to be writable for the docker_entrypoint.sh
+    chmod -f 0777 /opt/couchdb/etc/local.d
```
[the remark this answers is not in the document]
Comment (willholley) [reply 2.3.1/Dockerfile]: thanks @tianon - I've pushed another commit which does this. @wohali @kocolosk probably best to have another quick review before merging.

[the remark this answers is not in the document]
Comment (wohali) [reply 2.3.1/Dockerfile]: see his latest comment, looks like another change is req'd

[the remark this answers is not in the document]
Comment (willholley) [reply 2.3.1/Dockerfile]: ok - suggested change has been pushed

**Pull request merged**